dbms/src/Storages/DeltaMerge/File/ColumnCache.cpp

```diff
@@ -142,25 +142,68 @@ void ColumnCache::tryPutColumn(
     });
 }
 
-ColumnCacheElement ColumnCache::getColumn(size_t pack_id, ColId column_id)
+ColumnPtr ColumnCache::getColumn(size_t start_pack_id, size_t end_pack_id, size_t read_rows, ColId column_id)
+{
+    return column_caches.withShared([&](auto & column_caches) -> ColumnPtr {
+        auto iter = column_caches.find(start_pack_id);
+        RUNTIME_CHECK(iter != column_caches.end());
+        auto & columns = iter->second.columns;
+        auto col_iter = columns.find(column_id);
+        RUNTIME_CHECK_MSG(
+            col_iter != columns.end(),
+            "Cannot find column in cache for pack id: {}, column id: {}",
+            start_pack_id,
+            column_id);
+        const auto & column = col_iter->second;
+        // Optimization for some special cases:
+        // 1. The requested column is exactly the same as the cached column, return directly.
+        if (iter->second.rows_offset == 0 && column->size() == read_rows)
+            return column;
+        // 2. The requested column is a subset of the cached column, cut the cached column and return.
+        if (column->size() - iter->second.rows_offset >= read_rows)
+            return column->cut(iter->second.rows_offset, read_rows);
+
+        auto mut_col = column->cloneEmpty();
+        getColumn(mut_col, start_pack_id, end_pack_id, read_rows, column_id);
```

Comment:
Seems this will enter the following function at line 179 and acquire a shared_lock on `column_caches` again?

Reply:
fixed